core/src/main/scala/org/apache/spark/SparkContext.scala

```diff
@@ -1775,21 +1773,31 @@ class SparkContext(config: SparkConf) extends Logging {
     }
 
     val timestamp = if (addedOnSubmit) startTime else System.currentTimeMillis
+    // If the session ID was specified from SparkSession, it's from a Spark Connect client.
+    // Specify a dedicated directory for Spark Connect client.
+    // We're running Spark Connect as a service so regular PySpark path
+    // is not affected.
+    lazy val root = if (jobArtifactUUID != "default") {
+      val newDest = new File(SparkFiles.getRootDirectory(), jobArtifactUUID)
```

Comment:
Is this needed because the session-specific handling is more generic now? 
Because for the JARs from Spark Connect, we preiovusly just registered the root artifact directory in the file server and built URIs that let the executor fetch the file directly without the need of copying over to the generic Spark Files directory.

Reply:
Yeah, it now needs to reuse `PythonWorkerFactory` in which assumes that there is a UUID named directory under `SparkFiles.getRootDirectory()` at both Driver and Executor. We _could_ try to reuse the local artifact directory but I would prefer to have another copy in the local for now for better maintainability and reusability for now.

Otherwise, it does upload to the Spark file server twice (as we discussed offline). I pushed new changes to avoid this. So, after this change, we do not upload twice anymore by:
1. Directly pass the `spark://` URI to `addFile` and `addJar`
2. `addFile` and `addJar` will not attempt to upload the files, but bypass the original URI.

Reply:
>  reuse PythonWorkerFactory in which assumes that there is a UUID named directory under SparkFiles.getRootDirectory() at both Driver and Executor

Ahh gotcha, I am not very well aware of the Python side, good to know 👍 

> So, after this change, we do not upload twice anymore by:
Directly pass the spark:// URI to addFile and addJar
addFile and addJar will not attempt to upload the files, but bypass the original URI.

Awesome!